api/src/pcapi/core/offers/factories.py

```diff
@@ -63,8 +67,39 @@ def _create(
                 subcategory_id, str
             )  # if the subcategoryId was not given in the factory, it will get the default subcategoryId
             kwargs["extraData"] = build_extra_data_from_subcategory(
-                subcategory_id, kwargs.pop("set_all_fields", False), True
+                subcategory_id, kwargs.get("set_all_fields", False), True, is_offer=False
+            )
+
+        if "ean" not in kwargs:
+            fake = faker.Faker(locale="fr_FR")
+            subcategory_id = kwargs.get("subcategoryId")
+            assert isinstance(
+                subcategory_id, str
+            )  # if the subcategoryId was not given in the factory, it will get the default subcategoryId
+            subcategory = subcategories.ALL_SUBCATEGORIES_DICT.get(subcategory_id)
+            if not subcategory:
+                raise ValueError(f"Unknown subcategory {subcategory_id}")
+
+            conditional_fields = (
+                sorted(  # we sort to ensure MUSIC_SUB_TYPE and SHOW_SUB_TYPE is always after MUSIC_TYPE and SHOW_TYPE
+                    subcategory.conditional_fields,
+                    key=lambda field: (
+                        1
+                        if field
+                        in [
+                            subcategories.ExtraDataFieldEnum.MUSIC_SUB_TYPE.value,
+                            subcategories.ExtraDataFieldEnum.SHOW_SUB_TYPE.value,
+                        ]
+                        else 0
+                    ),
+                )
             )
```

Comment:
Ce morceau de code me parait compliqué pour juste dire "Ajouter un EAN si la sous-catégorie demande un EAN". Ce product est de catégorie Livres. Tous les products livres ont un EAN dans notre base, un peu plus largement, les products qui sont dans `THINGS_PRODUCT_SUBCATEGORIES_IDS` ont un EAN. 
Je retournerais ça en : 
```
if subcategoryId in THINGS_PRODUCT_SUBCATEGORIES_IDS:
    kwargs["ean"] = fake.ean13()
```
Ou même dans la définition un peu plus haut : 
`ean = factory.LazyAttribute(lambda o: fake.ean13() if subcategoryId in THINGS_PRODUCT_SUBCATEGORIES_IDS else None)`